### PR TITLE
Add XLA chunked KDA implementation and benchmark

### PR DESCRIPTION
**Pull request by Fred33146**
## Summary

- Add a pure-JAX/XLA chunked KDA implementation with separate forward and VJP entry points.
- Support fixed-length and packed/variable-length training inputs with chunk size 64.
- Register `xla_chunked` in the public KDA dispatch and benchmark matrix.
- Validate fixed and packed forward/VJP results against the recurrent XLA reference.

## Testing

- `python -m pytest -s -vv tokamax/_src/ops/experimental/kda/xla_chunked_test.py`
- `python -m tokamax.benchmarks.kda --skip_implementations=mosaic`

## Current limitations

- Only `chunk_size=64` is supported.
- Context parallelism is not yet supported.
- The backward path currently uses `jax.vjp` over the chunked forward implementation; a manually optimized backward implementation is follow-up work.

## Co-authors

Co-authored-by: Fred33146 liudonglai.ldl@antgroup.com\
Co-authored-by: fdz-1999 fengdaozhuo.fdz@antgroup.com

### CODE REVIEW

**Comment by RissyRan** [comment]
Will chat internally for the review triage.

**Comment by Majid-Taheri** [comment]
Thanks for submitting the Pr.
 In `xla_chunked_fwd_kernel.py` Lines 102 to 104 split the decay into two factors:

```
q_gated         = q_row * jnp.exp2(g_row - gate_reference)
k_inverse_gated = k_row * jnp.exp2(gate_reference - g_row)
```
The product is `exp2(g[i] - g[j])`, which stays in `(0, 1]` under the causal mask. But the second factor alone has a positive exponent. When the gate decays steeply it overflows to inf while the first underflows to 0, and the einsum on line 106 gives `inf * 0 = NaN`. Line 137 has the same shape for the off-diagonal block.

This probably does not show up in your tests. The gate there is 0.1 * normal and you run fp32 with HIGHEST precision. It might happens in bf16.

File: tokamax/_src/ops/experimental/kda/api.py [hunk 1]
```diff
@@ -20,12 +20,16 @@
...
 
 
-Implementation: TypeAlias = Literal["xla", "mosaic"]
+Implementation: TypeAlias = Literal["xla", "xla_chunked", "mosaic"]
 
-IMPLEMENTATIONS = dict(xla=base.KimiDeltaAttention())
+IMPLEMENTATIONS = dict(
+    xla=base.KimiDeltaAttention(),
+    xla_chunked=xla_chunked.XlaChunkedKimiDeltaAttention(),
+)
```
Comment (Majid-Taheri) [inline tokamax/_src/ops/experimental/kda/api.py:32]: Should xla_chunked sit in a try/except
like mosaic does?

**Comment by Fred33146** [comment]
Thanks for pointing this out. You are right that the factorized form is not numerically safe for an unbounded gate: one factor can underflow to `0` while the inverse factor overflows to `inf`, producing `inf * 0 = NaN`. The same theoretical issue also applied to the generic Pallas/Mosaic path.

These chunked backends are intended specifically as performance implementations aligned with our production XLA KDA configuration, where gate activation is fused and both the lower bound and safe-gate factoring are enabled. I have made that contract explicit in f1094f4 for both `xla_chunked` and `mosaic`, including their forward and VJP paths. They now require:

- `use_gate_in_kernel=True`
- a configured `lower_bound` (the API validates it to be in `[-5, 0)`)
- `safe_gate=True` after resolving its default value

Unsupported configurations now raise `NotImplementedError` before preprocessing or kernel tracing, so an ordered implementation list can fall back to the recurrent XLA implementation.

For the XLA chunk factorization mentioned here, each gate increment is therefore bounded to `[-5, 0]`. With `BC=16` and the safe-gate midpoint reference, the largest positive base-2 exponent is bounded by `(BC / 2) * 5 / ln(2) ~= 57.7`, which is well below the FP32 `exp2` overflow threshold of approximately `128`. Gate factors are computed in FP32 even for BF16 inputs. For the off-diagonal blocks, `gate_reference - g_column <= 0`, so the inverse factor cannot overflow there either. Thus the `inf * 0` case cannot occur within the supported production configuration.

I also updated the API documentation and added tests covering unsupported-configuration rejection and fallback for both backends. The relevant suite passes with `51 passed, 9 subtests passed`.

Thanks again for catching this.

File: tokamax/_src/ops/experimental/kda/api.py (hunk 1 above)
Comment (Fred33146) [reply tokamax/_src/ops/experimental/kda/api.py:32]: Thanks for raising this. I would prefer to keep `xla_chunked` imported unconditionally here.

`xla_chunked` does not require an additional platform-specific package. Although it transitively imports `jax.experimental.pallas.tpu` through a shared helper module, that module is part of the base JAX Python package; importing it does not require a TPU or inspect the active hardware backend. Backend availability only matters when lowering or executing a Pallas kernel.

Therefore, an import-time failure for `xla_chunked` would more likely indicate an incompatible JAX version, a broken dependency installation, or an actual code issue. Wrapping it in `try/except ImportError` could silently hide such errors.

Runtime configurations that `xla_chunked` does not support are handled separately by raising `NotImplementedError`, allowing an implementation sequence to fall back to the next backend. For those reasons, I would keep this import unconditional.

**Review by Majid-Taheri — APPROVED** [review]
approved thank you for raising this pr

**Comment by Majid-Taheri** [comment]
@Fred33146 looks like the new guard is causing these CI failure. Since the existing tests also use `use_gate_in_kernel_=False`, can you let unsplit path through this check?

**Comment by Fred33146** [comment]
Thanks for catching this. The new Mosaic guard was too broad: it rejected the existing pre-activated/log-space gate path (`use_gate_in_kernel=False`), which is exactly what the failing TPU kernel tests exercise.

I pushed 880477c to restore that path in Mosaic. The `lower_bound` and `safe_gate=True` checks still apply when gate activation is fused inside the kernel; the new `xla_chunked` backend remains restricted to that bounded configuration. I also updated the API documentation and adapter tests. The local KDA suite passes (`51 passed, 9 subtests passed`); the TPU CI rerun will validate the kernel tests.

I also want to correct my earlier numerical-stability statement: pre-activated log-space gates are not automatically bounded, and `lower_bound` does not constrain them when activation occurs outside the kernel. Restoring the existing Mosaic path preserves its prior behavior, but it does not guarantee overflow safety for arbitrary pre-activated gate values. The bounded-gate argument applies to the fused activation configuration.
